### ISSUE TITLE
AOTAutograd support for torch.export targeting model *inference*

### ISSUE THREAD

**Issue by Adam27X** [pytorch/pytorch#153251]
### 🚀 The feature, motivation and pitch

I'm working on accelerating some latency-sensitive PyTorch models and ideally I would be using torch.export for this use case because I would like to leverage some hand written optimizations in C++. A unique aspect of some of the models that I work with is that they require computation of their backward pass at _inference_ time because the gradient of the model's output has a physical/scientific meaning. This requirement prevents me from using torch.export for this use case because I cannot obtain a backward pass graph from torch.export's output.

### Alternatives

I'm aware of torch.aot_export_module, although this API seems targeted at training. There are incoming gradients to my model's backward pass from computation that occurs after the model's forward pass prior to calling autograd. It's unclear to me if there's a way to feed those gradients into the output of torch.aot_export_module. Additionally, because I'm doing inference here it's pretty unnatural to have the model return a loss.

Because of the limitations of these techniques I currently use a combination of torch.compile and aot_autograd. That works reasonably well, but torch.export would be more appropriate for my use case (I don't want graph breaks and I want to leverage dynamic shapes). Additionally, I've found at least one example where I'm able to compile the forward pass with torch.export but run into a bug trying to use torch.compile and aot_autograd (https://github.com/pytorch/pytorch/issues/152473). I'd rather use torch.export because it feels more stable and clearly user-facing whereas some of these alternatives aren't well documented and receive less support.

I'm very open to suggestions here and I appreciate your time and effort.

cc @chauhang @penguinwu @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4

**Comment by ydwu4** [comment]
Can try:
```python
import torch

class MyModule(torch.nn.Module):
    def __init__(self, val):
        super().__init__()
        self.val = val

    def forward(self, x):
        # force require grad to be able to calculate the gradient
        x.requires_grad_(True)
        y = x.sin()
        grad = torch.ones_like(y)
        return torch.autograd.grad(y, x, grad)

inputs = (torch.randn(3, 3),)
mod = MyModule(2)
ep = torch.export.export(mod, inputs)
print(ep.graph)
```
which gives a graph:
```
graph():
    %x : [num_users=2] = placeholder[target=x]
    %sin : [num_users=1] = call_function[target=torch.ops.aten.sin.default](args = (%x,), kwargs = {})
    %ones_like : [num_users=1] = call_function[target=torch.ops.aten.ones_like.default](args = (%sin,), kwargs = {pin_memory: False})
    %cos : [num_users=1] = call_function[target=torch.ops.aten.cos.default](args = (%x,), kwargs = {})
    %mul : [num_users=1] = call_function[target=torch.ops.aten.mul.Tensor](args = (%ones_like, %cos), kwargs = {})
    return (mul,)
```
Is this what you need?

**Comment by Adam27X** [comment]
Thanks for the quick response. The flow I'm working with looks something like the following:

```
class MyModule(torch.nn.Module):
    def __init__(self, val):
        super().__init__()
        self.val = val

    def forward(self, x):
        # force require grad to be able to calculate the gradient
        x.requires_grad_(True)
        y = x.sin()
        return y

def iter(mod, x):
  y = mod.forward(x)
  z = 2*y #Simple example of compute outside of the model, but also consider something like a call to a C++ library here
  return torch.autograd.grad(z.sum(), x)
```

With the exception that there are multiple grads from the computation after `mod.forward()` that need to flow into `mod.backward()`.

Additionally, when I tried to run your example (using PyTorch 2.7.0) I hit

```
torch._dynamo.exc.Unsupported: Tensor.requires_grad_
from user code:
   File "/u/pgh/mclaugha/./sandbox/issue_153251/test.py", line 10, in forward
    x.requires_grad_(True)
```

but then if I try to remove that from the function itself and simply pass in an input that requires grad I instead get:

```
torch._dynamo.exc.Unsupported: Attempted to call function marked as skipped
  Explanation: Dynamo developers have intentionally marked that the function `grad` in file `/u/pgh/mclaugha/.venv/lib/python3.10/site-packages/torch/autograd/__init__.py` should not be traced.
  Hint: Avoid calling the function `grad`.
  Hint: Remove the function `grad` or the file `/u/pgh/mclaugha/.venv/lib/python3.10/site-packages/torch/autograd/__init__.py` from torch/_dynamo/trace_rules.py. More graph breaks may occur as a result of attempting to trace into the function.
  Hint: Please file an issue to PyTorch.
  Developer debug context: module: torch.autograd, qualname: grad, skip reason: <missing reason>
from user code:
   File "/u/pgh/mclaugha/./sandbox/issue_153251/test.py", line 11, in forward
    return torch.autograd.grad(y, x, grad)
```

Same result if I try `y.backward()` instead.

**Comment by ydwu4** [comment]
Can you try nightly or pass in a strict=False flag to the export call? something like torch.expor.export(...strict=False). strict=False is now the default. It skips dynamo.

**Comment by Adam27X** [comment]
> Can you try nightly or pass in a strict=False flag to the export call? something like torch.expor.export(...strict=False). strict=False is now the default. It skips dynamo.

This gets the example working for me, thank you!

I suppose the best way to move forward with my real use case then is to wrap the entire computation into a new `torch.nn.Module` instance where all of the computation occurs within that new model's `forward` function and to use `torch.library.custom_op` to register any C++ utilities so they appear in the FxGraph. Does that sound reasonable?

**Comment by ydwu4** [comment]
Yeah, that sounds reasonable. Happy hacking!

**Comment by Adam27X** [comment]
Thanks again! Closing.
